wagtail/admin/views/generic/base.py

```diff
@@ -113,3 +120,199 @@ def post(self, request, *args, **kwargs):
+        # PK if not (to ensure pagination is consistent)
+        if not queryset.ordered:
+            queryset = queryset.order_by("pk")
+
+        return queryset
+
+    def paginate_queryset(self, queryset, page_size):
+        paginator = self.get_paginator(
+            queryset,
+            page_size,
+            orphans=self.get_paginate_orphans(),
+            allow_empty_first_page=self.get_allow_empty(),
+        )
+
+        page_number = self.request.GET.get(self.page_kwarg)
+        page = paginator.get_page(page_number)
+        return (paginator, page, page.object_list, page.has_other_pages())
+
+    def get_filterset_class(self):
+        if self.filterset_class:
+            return self.filterset_class
+
+        if not self.list_filter or not self.model:
+            return None
+
+        class Meta:
+            model = self.model
+            fields = self.list_filter
+
+        return type(
+            f"{self.model.__name__}FilterSet",
+            (WagtailFilterSet,),
+            {"Meta": Meta},
+        )
+
+    def filter_queryset(self, queryset):
+        # construct filter instance (self.filters) if not created already
+        if self.filterset_class and self.filters is None:
+            self.filters = self.filterset_class(
+                self.request.GET, queryset=queryset, request=self.request
+            )
+            queryset = self.filters.qs
+        elif self.filters:
+            # if filter object was created on a previous filter_queryset call, re-use it
+            queryset = self.filters.filter_queryset(queryset)
+
+        return self.filters, queryset
+
+    def search_queryset(self, queryset):
+        if not self.search_query:
+            return queryset
+
+        if class_is_indexed(queryset.model) and self.search_backend_name:
+            search_backend = get_search_backend(self.search_backend_name)
+            if self.use_autocomplete:
+                return search_backend.autocomplete(
+                    self.search_query, queryset, fields=self.search_fields
+                )
+            return search_backend.search(
+                self.search_query, queryset, fields=self.search_fields
+            )
+
+        filters = {
+            field + "__icontains": self.search_query
+            for field in self.search_fields or []
+        }
+        return queryset.filter(**filters)
+
+    def get_valid_orderings(self):
+        orderings = []
+        for col in self.columns:
+            if col.sort_key:
+                orderings.append(col.sort_key)
+                orderings.append("-%s" % col.sort_key)
+        return orderings
+
+    def get_ordering(self):
+        ordering = self.request.GET.get("ordering", self.default_ordering)
+        if ordering not in self.get_valid_orderings():
+            ordering = self.default_ordering
+        return ordering
+
+    def get_table(self, object_list, **kwargs):
+        return self.table_class(
+            self.columns,
```

Comment:
Shall we add an empty `columns = []` on the class, and/or a simple implementation of `get_columns()`? There are references to `self.columns` in this class but it has no assignment to that variable.

Reply:
Have now added a default columns definition that just displays `__str__` in https://github.com/gasman/wagtail/commit/7324cb0f2362b454e5df528589ca71acb774ff27. Was originally thinking of this as an abstract base class where subclasses would have to specify the columns in order for it to be useful (since this can be a listing of _anything_, not necessarily a model), but `__str__` works as a suitably generic default I think.

Reply:
Oops... looks like that breaks the logic in generic.models.IndexView, which only falls back on generating the columns from `list_display` if the class-level `columns` hasn't been set. Needs more thought...

Reply:
Yeah I think using `__str__` makes sense!